articles/actions.md

```diff
@@ -0,0 +1,338 @@
+They allow a client to track the progress of a request, get the final outcome, and optionally cancel the request it before it completes.
+
+This document defines requirements for actions, how they are specified in the ROS Message IDL, and how they are communicated by the middleware.
+
+## Entities Involved in Actions
+
+There are two entities involved in actions: an action server and an action client.
+
+### Action Server
+
+An action server provides an action.
+Like topics and services, an action server has a name and a type.
+There is only one server per name.
+
+It is responsible for
+
+* advertising the action to other ROS entities
+* accepting or rejecting requests from one or more action clients
+* executing the action when a request is received and accepted
+* optionally providing feedback about the progress of all executing actions
+* optionally handling requests to cancel an action
+* sending the result of the action, including whether it succeed, failed, or was cancelled, to the client when the action completes
+
+### Action Client
+
+An action client requests an action to be performed and monitors its progress.
+There may be multiple clients per server; however, it is up to the server to decide how requests from multiple clients will be handled.
+
+It is responsible for
+
+* making a request to the action server
+* optionally monitoring the feedback from the action server
+* optionally requesting that the action server cancel the action execution
+* optionally checking the result of the action received from the action server
+
+## Differences between ROS 1 and ROS 2 actions
+
+### First Class Support
+In ROS 1, actions are implemented as a separate library `actionlib` which is built on top of the client libraries.
+This was done to avoid increasing the work required to create a client library in a new language, but actions turned out to be very important to packages like the navigation stack<sup>[1](#separatelib)</sup>.
+In ROS 2 actions will be included in the client library implementations.
+The work of writing a client library in a new language will be reduced by creating a common implmentation in C.
+
+### Services used for Actions
+
+In ROS 1 actions were implemented using a set of topics under a namespace taken from the action name.
+ROS 1 Services were not used because they are inherently synchronous, and actions need to be asynchronous.
+Actions also needed to send status/feedback and be cancelable.
+In ROS 2 services are asynchronous in the common C implementation, so actions will use a combination of services and topics.
+
+### Goal Identifiers
+
+In ROS 1, Action clients are responsible for creating a goal ID when submitting a goal.
+In ROS 2 the action server will be responsible for generating the goal ID and notifying the client.
+
+The server is better equiped to generate a unique goal id than the client because there may be multiple clients who could independently generate the same goal id.
```

Comment:
using a UUID may mitigate this, and still allow the client to 'know' the ID of a goal without ever talking to a server. This can be helpful when correlating 'actions' with other libraries/systems.

Reply:
This seems reasonable. The original reason for having the server generate the ID is to avoid a race condition where the client may try to use the ID to cancel a goal before it is accepted. But in retrospect, by including the feature to "cancel all goals" or "cancel all goals before timestamp" we can still have a scenario where the cancel request arrives before a goal request. I guess this scenario can be resolved by processing requests as they arrive at the server; if a cancel request arrives first, then any subsequent goal requests are not retroactively canceled. Clients will still be informed that the goal is accepted via the goal response.

In the event that a client tries to use the explicit ID to cancel a goal before it is accepted, the server can reject the request if there are no goals with that ID in a valid state (PENDING or EXECUTING).

Therefore, I see no issue with allowing clients to generate a UUID.

Reply:
You used the name "PENDING/EXECUTING" here -- I would suggest we actually use "ACCEPTED" and "EXECUTING" as the two state names -- it really tells you what stage the goal is in (accepted seems to have a better semantic meaning than pending, pending could mean that handle_goal() has yet to be called or something else. accepted really does state the the action server has acknowledged the goal, but not starting executing it).

Reply:
> But in retrospect, by including the feature to "cancel all goals" or "cancel all goals before timestamp" we can still have a scenario where the cancel request arrives before a goal request. I guess this scenario can be resolved by processing requests as they arrive at the server; if a cancel request arrives first, then any subsequent goal requests are not retroactively canceled

I wouldn't expect a goal to be cancellable until it has been accepted. Even if the goal is cancelled between the goal acceptance being transmitted from server to client then the client would still receive a response to its goal request indicating the goal was accepted, followed by a response to its result request indicating the goal was cancelled.

I guess the real problem in ROS 1 isn't the goal id generation, but that `actionlib` doesn't notify the client when a cancel request fails. It seems like the race condition is actually solved by cancel becoming a service. 

> using a UUID may mitigate this, and still allow the client to 'know' the ID of a goal without ever talking to a server. This can be helpful when correlating 'actions' with other libraries/systems.

A UUID generated by a client would work.